src/libraries/System.Text.RegularExpressions/src/System/Text/RegularExpressions/RegexInterpreter.cs

```diff
@@ -207,6 +207,18 @@ private void StackPush(int i1, int i2)
 
         private int Forwardchars() => _rightToLeft ? runtextpos - runtextbeg : runtextend - runtextpos;
 
+        private bool IsBoundary(ReadOnlySpan<char> inputSpan, int index, int startpos, int endpos)
+        {
+            return (index > startpos && RegexCharClass.IsBoundaryWordChar(inputSpan[index - 1])) !=
+                   (index < endpos && RegexCharClass.IsBoundaryWordChar(inputSpan[index]));
+        }
+
+        private bool IsECMABoundary(ReadOnlySpan<char> inputSpan, int index, int startpos, int endpos)
+        {
+            return (index > startpos && RegexCharClass.IsECMAWordChar(inputSpan[index - 1])) !=
+                   (index < endpos && RegexCharClass.IsECMAWordChar(inputSpan[index]));
+        }
```

Comment:
We shouldn't need the same helpers in multiple places.  You can just put it as internal somewhere and use it from everywhere we need it (except for the source generator).  The existing string-based methods should be able to just delegate to the span-based ones.